src/test/spec/v2_runner/mod.rs

```diff
@@ -232,220 +226,282 @@ async fn run_v2_test(path: std::path::PathBuf, test_file: TestFile) {
+    ) -> Option<Result<Option<bson::Bson>, crate::error::Error>> {
+        if operation.name == "endSession" {
+            let session = match &operation.object {
+                Some(OperationObject::Session0) => &mut self.session0,
+                Some(OperationObject::Session1) => &mut self.session1,
+                other => panic!("invalid object for `endSession`: {:?}", other),
             };
+            drop(session.take());
+            runtime::delay_for(Duration::from_secs(1)).await;
+            return None;
+        }
 
-            let result = match operation.object {
-                Some(OperationObject::Collection) | None => {
-                    let result = operation.execute_on_collection(&coll, session).await;
-                    // This test (in src/test/spec/json/sessions/server-support.json) runs two
-                    // operations with implicit sessions in sequence and then checks to see if they
-                    // used the same lsid. We delay for one second to ensure that the
-                    // implicit session used in the first operation is returned to the pool before
-                    // the second operation is executed.
-                    if test.description == "Server supports implicit sessions" {
-                        runtime::delay_for(Duration::from_secs(1)).await;
-                    }
-                    result
-                }
-                Some(OperationObject::Database) => {
-                    operation.execute_on_database(&db, session).await
+        let sessions = OpSessions {
+            session0: self.session0.as_mut(),
+            session1: self.session1.as_mut(),
+        };
+
+        let mut runner = OpRunner {
+            description: self.description.clone(),
+            internal_client: self.internal_client.clone(),
+            client: self.client.clone(),
+            ns: self.ns.clone(),
+            fail_point_guards: &mut self.fail_point_guards,
+        };
+
+        runner.run_operation(operation, sessions).await
+    }
+}
+
+pub(crate) struct OpSessions<'a> {
+    session0: Option<&'a mut ClientSession>,
+    session1: Option<&'a mut ClientSession>,
+}
+
+pub(crate) struct OpRunner<'a> {
+    description: String,
+    internal_client: TestClient,
+    client: EventClient,
+    ns: Namespace,
+    fail_point_guards: &'a mut Vec<FailPointGuard>,
+}
+
+impl<'a> OpRunner<'a> {
+    pub(crate) async fn run_operation<'b>(
+        &mut self,
+        operation: &Operation,
+        mut sessions: OpSessions<'b>,
+    ) -> Option<Result<Option<bson::Bson>, crate::error::Error>> {
+        if operation.name == "withTransaction" {
+            if !matches!(&operation.object, Some(OperationObject::Session0)) {
+                panic!("invalid object for withTransaction: {:?}", operation.object);
+            }
+            return Some(operation.execute_recursive(self, sessions).await);
+        }
+
+        let db = match &operation.database_options {
```

Comment:
Despite what github's diff says, the rest of the logic in this method is unchanged from the original loop.